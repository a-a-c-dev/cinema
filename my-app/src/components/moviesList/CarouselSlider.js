import React, {useState, useEffect, useMemo, Suspense, lazy} from "react";
import {CarouselLeftArrow} from "./CarouselLeftArrow";
import {CarouselRightArrow} from "./CarouselRightArrow";
import {Spinner} from '../common/Spinner';


const MoviesContainer = lazy(() => import("./MoviesContainer"));


const CarouselSlider = ({ movies =[], handleEdit,handleInfo, handleDelete }) => {
  const [indexSlider, setIndexSlider] = useState(1);
  const [prevIndexSlider, setPrevIndexSlider] = useState(movies.length-2);
  const [nextIndexSlider, setNextIndexSlider] = useState(2);
  const [activeIndexSlider, setActiveIndexSlider] = useState(1);  

  const prevMovie = useMemo(() => event => {
    event.preventDefault();
    let lastIndex = movies.length-1;
    (indexSlider-1 >= 0)?setIndexSlider(indexSlider-1) :setIndexSlider(lastIndex);
    ( activeIndexSlider-1<0)?setPrevIndexSlider(currentMovies.length-1):setPrevIndexSlider(activeIndexSlider-1);
    ( activeIndexSlider+1>currentMovies.length-1)?setNextIndexSlider(0):setNextIndexSlider(activeIndexSlider+1);
  },[indexSlider]);

  const nextMovie =useMemo(()=> event => {
    event.preventDefault();
    console.log("nextIndexSlider",nextIndexSlider)
    let lastIndex = movies.length-1;
    (indexSlider+1 > lastIndex)? setIndexSlider(0): setIndexSlider(indexSlider+1);
    ( activeIndexSlider-1<0)?setPrevIndexSlider(currentMovies.length-1):setPrevIndexSlider(activeIndexSlider-1);
    ( activeIndexSlider+1>currentMovies.length-1)?setNextIndexSlider(0):setNextIndexSlider(activeIndexSlider+1);
  },[indexSlider]);

const currentMovies = useMemo(()=>movies,[movies]);

const handleMemoEdit = useMemo(()=> handleEdit, [])
const handleMemoInfo = useMemo(()=> handleInfo, [])
const handleMemoDelete = useMemo(()=> handleDelete, [])

useEffect(() => {
  setActiveIndexSlider(indexSlider);
}, [indexSlider]); 

useEffect(() => {
  setActiveIndexSlider(movies.length-1);
}, [movies]); 

useEffect(() => {
  ( activeIndexSlider-1<0)?setPrevIndexSlider(currentMovies.length-1):setPrevIndexSlider(activeIndexSlider-1);
  ( activeIndexSlider+1>currentMovies.length-1)?setNextIndexSlider(0):setNextIndexSlider(activeIndexSlider+1);
}, [activeIndexSlider]); 


  return (
      <div className="carousel-slider" >
        <CarouselRightArrow nextMovie={nextMovie}/>
        <Suspense fallback={<Spinner/>}>
          <MoviesContainer 
              currentMovies={currentMovies}  
              activeIndexSlider={activeIndexSlider}
              prevIndexSlider={prevIndexSlider}
              nextIndexSlider={nextIndexSlider}
              handleEdit={handleMemoEdit}
              handleDelete={handleMemoDelete}
              handleInfo={handleMemoInfo}
          />
        </Suspense>
        <CarouselLeftArrow prevMovie={prevMovie}/> 
    </div>
  );
};

export default CarouselSlider;



