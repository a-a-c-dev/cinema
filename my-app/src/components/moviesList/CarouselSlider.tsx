import React, {useState, useEffect, useMemo, Suspense, lazy} from "react";
import {CarouselLeftArrow} from "./CarouselLeftArrow";
import {CarouselRightArrow} from "./CarouselRightArrow";
import {Spinner} from '../common/Spinner';
import {Movie} from '../../App'


const MoviesContainer = lazy(() => import("./MoviesContainer"));

interface CarouselSliderProps{
  movies :Movie[],
  handleEdit:(movie:Movie)=>void,
  handleInfo:(movie:Movie)=>void,
  handleDelete:(movie:Movie)=>void
}

const CarouselSlider = ({ movies =[], handleEdit,handleInfo, handleDelete }:CarouselSliderProps) => {
  const [indexSlider, setIndexSlider] = useState<number>(1);
  const [prevIndexSlider, setPrevIndexSlider] = useState<number>(movies.length-2);
  const [nextIndexSlider, setNextIndexSlider] = useState<number>(2);
  const [activeIndexSlider, setActiveIndexSlider] = useState<number>(1);  

  const prevMovie = useMemo(() => (event:React.MouseEvent<HTMLButtonElement>) => {
    event.preventDefault();
    let lastIndex = movies.length-1;
    (indexSlider-1 >= 0)?setIndexSlider(indexSlider-1) :setIndexSlider(lastIndex);
    ( activeIndexSlider-1<0)?setPrevIndexSlider(currentMovies.length-1):setPrevIndexSlider(activeIndexSlider-1);
    ( activeIndexSlider+1>currentMovies.length-1)?setNextIndexSlider(0):setNextIndexSlider(activeIndexSlider+1);
  },[indexSlider]);

  const nextMovie =useMemo(()=> (event:React.MouseEvent<HTMLButtonElement>) => {
    event.preventDefault();
    let lastIndex = movies.length-1;
    (indexSlider+1 > lastIndex)? setIndexSlider(0): setIndexSlider(indexSlider+1);
    ( activeIndexSlider-1<0)?setPrevIndexSlider(currentMovies.length-1):setPrevIndexSlider(activeIndexSlider-1);
    ( activeIndexSlider+1>currentMovies.length-1)?setNextIndexSlider(0):setNextIndexSlider(activeIndexSlider+1);
  },[indexSlider]);

const currentMovies = useMemo(()=>movies,[movies]);

const handleMemoEdit = useMemo(()=> handleEdit, [])
const handleMemoInfo = useMemo(()=> handleInfo, [])
const handleMemoDelete = useMemo(()=> handleDelete, [])

useEffect(() => {
  setActiveIndexSlider(indexSlider);
}, [indexSlider]); 

useEffect(() => {
  setActiveIndexSlider(movies.length-1);
}, [movies]); 

useEffect(() => {
  ( activeIndexSlider-1<0)?setPrevIndexSlider(currentMovies.length-1):setPrevIndexSlider(activeIndexSlider-1);
  ( activeIndexSlider+1>currentMovies.length-1)?setNextIndexSlider(0):setNextIndexSlider(activeIndexSlider+1);
}, [activeIndexSlider]); 


  return (
      <div className="carousel-slider" >
        <CarouselRightArrow nextMovie={nextMovie}/>
        <Suspense fallback={<Spinner/>}>
          <MoviesContainer 
              currentMovies={currentMovies}  
              activeIndexSlider={activeIndexSlider}
              prevIndexSlider={prevIndexSlider}
              nextIndexSlider={nextIndexSlider}
              handleEdit={handleMemoEdit}
              handleDelete={handleMemoDelete}
              handleInfo={handleMemoInfo}
          />
        </Suspense>
        <CarouselLeftArrow prevMovie={prevMovie}/> 
    </div>
  );
};

export default CarouselSlider;



