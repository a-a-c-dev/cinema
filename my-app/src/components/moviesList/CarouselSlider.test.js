import React from 'react';
import { shallow } from 'enzyme';
import  CarouselSlider from './CarouselSlider';


describe('List component', () => {
    it('should render correctly in "debug" mode', () => {
        const component = shallow(<CarouselSlider debug />);
        expect(component).toMatchSnapshot();
    });
});