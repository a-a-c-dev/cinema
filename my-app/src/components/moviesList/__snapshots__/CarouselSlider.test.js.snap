// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List component should render correctly in "debug" mode 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CarouselSlider
    debug={true}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <CarouselRightArrow
          nextMovie={[Function]}
        />,
        <UNDEFINED
          fallback={<Spinner />}
        >
          <UNDEFINED
            activeIndexSlider={1}
            currentMovies={Array []}
            nextIndexSlider={2}
            prevIndexSlider={-2}
          />
        </UNDEFINED>,
        <CarouselLeftArrow
          prevMovie={[Function]}
        />,
      ],
      "className": "carousel-slider",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "nextMovie": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <UNDEFINED
            activeIndexSlider={1}
            currentMovies={Array []}
            nextIndexSlider={2}
            prevIndexSlider={-2}
          />,
          "fallback": <Spinner />,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "activeIndexSlider": 1,
            "currentMovies": Array [],
            "handleDelete": undefined,
            "handleEdit": undefined,
            "handleInfo": undefined,
            "nextIndexSlider": 2,
            "prevIndexSlider": -2,
          },
          "ref": null,
          "rendered": null,
          "type": Object {
            "$$typeof": Symbol(react.lazy),
            "_ctor": [Function],
            "_result": null,
            "_status": -1,
          },
        },
        "type": Symbol(react.suspense),
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "prevMovie": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <CarouselRightArrow
            nextMovie={[Function]}
          />,
          <UNDEFINED
            fallback={<Spinner />}
          >
            <UNDEFINED
              activeIndexSlider={1}
              currentMovies={Array []}
              nextIndexSlider={2}
              prevIndexSlider={-2}
            />
          </UNDEFINED>,
          <CarouselLeftArrow
            prevMovie={[Function]}
          />,
        ],
        "className": "carousel-slider",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "nextMovie": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <UNDEFINED
              activeIndexSlider={1}
              currentMovies={Array []}
              nextIndexSlider={2}
              prevIndexSlider={-2}
            />,
            "fallback": <Spinner />,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "activeIndexSlider": 1,
              "currentMovies": Array [],
              "handleDelete": undefined,
              "handleEdit": undefined,
              "handleInfo": undefined,
              "nextIndexSlider": 2,
              "prevIndexSlider": -2,
            },
            "ref": null,
            "rendered": null,
            "type": Object {
              "$$typeof": Symbol(react.lazy),
              "_ctor": [Function],
              "_result": null,
              "_status": -1,
            },
          },
          "type": Symbol(react.suspense),
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "prevMovie": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
}
`;
